### PR TITLE
Pulselistener fixes

### PR DESCRIPTION
**Pull request by La0**
## Part 1: Update mercurial to 5.1 
... and use up-to-date version-control-tools

## Part 2 : Support 3 clone modes, configurable through TC secret

* Batch_checkout (as before)
* Robustcheckout
* Default Hg clone
This change is not mandatory, as i made it to debug pulselistener crashes on a day where the HG payload on the cdn was corrupted (when i understood the issue, i reported it on #vcs, it was fixed in a few minutes…)
But, having robustcheckout is nice to have for smaller repos like nss

## Part 3: Redis support in Message bus
It’s only used for the incoming HTTP notifications. As it’s limited in scope, it was simple to implement (dump the pickled objects in a redis list). But the benefits are immediate: even if pulselistener crashes, we do not lose anymore the build notifications from the previous instances

## Part 4: Fix the end of life cycle.

The web process was not stopped on crashes (only on KeyboardInterrupt…) I simplified the old run_consumer function from release-services into a run_tasks that use a cancellable task (still using asyncio.gather) and avoiding the run_forever. Basically the program now can stop cleanly. This allows Heroku to restart immediately pulselistener instead of waiting without doing nothing

## Part 5: Fix a nasty bug in the Taskcluster lib usage for code coverage

the limit & continuationToken were passed as kwargs and ignored by the lib (instead we need to provice them as a query).
The function was always returning the same set of tasks on the first page, which is fine in most cases where the number of tasks is below the limit (1000), but on those rare task group with more than 1000 tasks, it was locking the whole program and always retrieving that first page.
While debugging this, i found out that the code coverage part of pulselistener was spending way more time than i thought retrieving those Taskcluster payloads. So i cleaned up the code around that part, added some logging, a better retry, and made it async to allow a small asyncio.sleep so it does not lock other processes while working on big batches.

## Part 6: Simplification of the pulse connection logic

In order to remove an extra useless while True (useless for us because we already run that code in a while True).

### CODE REVIEW

**Comment by La0** [comment]
Running on testing, with `robust` enabled, got a treeherde link !

**Comment by La0** [comment]
I deployed the shutdown management patch on Heroku's testing env: the worker now stops cleanly when it encounters a clone issue.

**Comment by La0** [comment]
Pulselistener now fully works again on testing.

**Comment by La0** [comment]
That code is now running in testing (for a full day) and on production since one hour.

**Comment by La0** [comment]
Cleaned up commit history to match doc.

File: src/pulselistener/pulselistener/lib/mercurial.py [hunk 1]
```diff
@@ -72,17 +72,15 @@ def end_of_life(self):
...
-        loop = asyncio.get_running_loop()
-        with ProcessPoolExecutor() as pool:
-            logger.info('Checking out tip in a separate process', repo=self.url)
-            await loop.run_in_executor(
-                pool,
-                batch_checkout,
-                self.url, self.dir, b'tip', self.batch_size,
-            )
-            logger.info('Batch checkout finished')
+    def clone(self):
```
Comment (marco-c) [inline src/pulselistener/pulselistener/lib/mercurial.py]: Shouldn't we keep cloning in a separate process to avoid blocking other running tasks?

File: src/pulselistener/pulselistener/code_review.py [hunk 2]
```diff
@@ -57,6 +57,8 @@ def get_repositories(self, repositories, cache_root):
 
             # Receive build from webserver
             build = await self.bus.receive(QUEUE_WEB_BUILDS)
+            if build is None:
```
Comment (marco-c) [inline src/pulselistener/pulselistener/code_review.py]: When does this happen? Do we risk entering in an infinite loop?

File: src/pulselistener/pulselistener/lib/bus.py [hunk 3]
```diff
@@ -35,8 +59,15 @@ def add_queue(self, name, mp=False, maxsize=-1):
         '''
         assert name in self.queues, 'Missing queue {}'.format(name)
         queue = self.queues[name]
-        if isinstance(queue, asyncio.Queue):
+
+        if isinstance(queue, RedisQueue):
+            async with AsyncRedis() as redis:
```
Comment (marco-c) [inline src/pulselistener/pulselistener/lib/bus.py]: Maybe it'd be better to connect to redis only once instead of each time we send a message

File: src/pulselistener/pulselistener/lib/bus.py [hunk 4]
```diff
@@ -48,8 +79,24 @@ def add_queue(self, name, mp=False, maxsize=-1):
         '''
         assert name in self.queues, 'Missing queue {}'.format(name)
         queue = self.queues[name]
-        if isinstance(queue, asyncio.Queue):
+
+        logger.debug('Wait for message on bus', queue=name, instance=queue)
+
+        if isinstance(queue, RedisQueue):
+            async with AsyncRedis() as redis:
```
Comment (marco-c) [inline src/pulselistener/pulselistener/lib/bus.py]: Same here

File: src/pulselistener/pulselistener/lib/bus.py [hunk 5]
```diff
@@ -35,8 +59,15 @@ def add_queue(self, name, mp=False, maxsize=-1):
         '''
         assert name in self.queues, 'Missing queue {}'.format(name)
         queue = self.queues[name]
-        if isinstance(queue, asyncio.Queue):
+
+        if isinstance(queue, RedisQueue):
+            async with AsyncRedis() as redis:
+                nb = await redis.rpush(queue.name, pickle.dumps(payload))
```
Comment (marco-c) [inline src/pulselistener/pulselistener/lib/bus.py]: Nit: maybe we can use HIGHEST_PROTOCOL to have the best protocol

File: src/pulselistener/pulselistener/lib/bus.py [hunk 6]
```diff
@@ -99,7 +100,15 @@ def add_queue(self, name, mp=False, redis=False, maxsize=-1):
 
         else:
             # Run the synchronous mp queue.get in the asynchronous loop
-            return await asyncio.get_running_loop().run_in_executor(None, queue.get)
+            # but use an asyncio sleep to be able to react to cancellation
```
Comment (marco-c) [inline src/pulselistener/pulselistener/lib/bus.py]: I wonder if there isn't a nicer way to achieve this

File: src/pulselistener/pulselistener/lib/utils.py [hunk 7]
```diff
@@ -35,6 +36,30 @@ def retry(operation,
...
+    Helper to run tasks concurrenlty, but when an exception is raised
+    by one of the tasks, the whole stack stops.
+    '''
+    assert isinstance(awaitables, list)
+
+    async def _run():
+        try:
+            # Create a task grouping all awaitables
+            # and running them concurrently
+            task = asyncio.gather(*awaitables)
```
Comment (marco-c) [inline src/pulselistener/pulselistener/lib/utils.py]: Nit: `tasks_future`

File: src/pulselistener/pulselistener/lib/utils.py [hunk 8]
```diff
@@ -116,7 +141,7 @@ def batch_checkout(repo_url, repo_dir, revision=b'tip', batch_size=100000):
 
 def robust_checkout(repo_url, repo_dir, branch=b'tip'):
     '''
-    Helper to clone Mozilla Central using the robustcheckout extension
+    Helper to clone a mercurial repo using the robustcheckout extension
```
Comment (marco-c) [inline src/pulselistener/pulselistener/lib/utils.py]: Nit: I guess this change belongs to another commit :P

File: src/pulselistener/pulselistener/lib/utils.py [hunk 9]
```diff
@@ -35,6 +36,30 @@ def retry(operation,
             time.sleep(wait_between_retries)
 
 
+def run_tasks(awaitables):
+    '''
+    Helper to run tasks concurrenlty, but when an exception is raised
```
Comment (marco-c) [inline src/pulselistener/pulselistener/lib/utils.py]: Typo: concurrently

File: src/pulselistener/pulselistener/listener.py [hunk 10]
```diff
@@ -60,48 +62,52 @@ def __init__(self, configuration, bus):
...
-                )
-                task = maybe_trigger(reply['tasks'])
+        async def retrieve_coverage_task():
+            task, token = load_tasks()
+
+            while task is None and token is not None:
+                task, token = load_tasks(continuationToken=token)
+
+                # Let other tasks run on long batches
+                await asyncio.sleep(2)
```
Comment (marco-c) [inline src/pulselistener/pulselistener/listener.py]: You can use sleep(0) to allow other tasks to run.

(It would be even better to use the async version of the Taskcluster lib, not sure how easy it is though)

File: src/pulselistener/pulselistener/lib/pulse.py [hunk 11]
```diff
@@ -123,6 +130,7 @@ def register(self, bus):
         body = json.loads(body.decode('utf-8'))
 
         # Push the message in the message bus
+        logger.debug('Received a pulse message')
```
Comment (marco-c) [inline src/pulselistener/pulselistener/lib/pulse.py]: Nit: Maybe we should avoid logging this, as we get too many messages

Comment (La0) [reply src/pulselistener/pulselistener/lib/pulse.py]: In debug it's never displayed by default, only when using a verbose mode while developing

File: src/pulselistener/pulselistener/lib/mercurial.py (hunk 1 above)
Comment (La0) [reply src/pulselistener/pulselistener/lib/mercurial.py]: This is one step ahead, but when running only the code-review workflow, it's useless to have other processes running before the mercurial checkout is done (except the web process which is in its own process already).

I can rollback to split processes for that part if you want

File: src/pulselistener/pulselistener/code_review.py (hunk 2 above)
Comment (La0) [reply src/pulselistener/pulselistener/code_review.py]: This happen when i do `rpush webserver:out "THIS IS A CRASH"` :smile: 

When i was testing the integration, it was immediately crashing the whole program, so i prefer to have a solution for bad data...

File: src/pulselistener/pulselistener/lib/bus.py (hunk 3 above)
Comment (La0) [reply src/pulselistener/pulselistener/lib/bus.py]: Ok, will do.

File: src/pulselistener/pulselistener/lib/utils.py (hunk 8 above)
Comment (La0) [reply src/pulselistener/pulselistener/lib/utils.py]: Oops.

File: src/pulselistener/pulselistener/listener.py (hunk 10 above)
Comment (La0) [reply src/pulselistener/pulselistener/listener.py]: I tried to use the async lib... and failed (not given much time). It's there as `aio` in the tc lib; but no example

File: src/pulselistener/pulselistener/lib/bus.py (hunk 6 above)
Comment (La0) [reply src/pulselistener/pulselistener/lib/bus.py]: I can rollback to the initial code here, as it is replaced by Redis queue on Heroku (and even dev)

File: src/pulselistener/pulselistener/lib/mercurial.py (hunk 1 above)
Comment (marco-c) [reply src/pulselistener/pulselistener/lib/mercurial.py]: OK, feel free to leave it as it is.

File: src/pulselistener/pulselistener/code_review.py (hunk 2 above)
Comment (marco-c) [reply src/pulselistener/pulselistener/code_review.py]: Maybe we should assert instead? If we have a stream of bad data we will loop forever otherwise

File: src/pulselistener/pulselistener/lib/bus.py (hunk 3 above)
Comment (La0) [reply src/pulselistener/pulselistener/lib/bus.py]: Done.

File: src/pulselistener/pulselistener/lib/bus.py (hunk 4 above)
Comment (La0) [reply src/pulselistener/pulselistener/lib/bus.py]: Done.

File: src/pulselistener/pulselistener/code_review.py (hunk 2 above)
Comment (La0) [reply src/pulselistener/pulselistener/code_review.py]: Done.
